source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h

```diff
@@ -294,7 +301,14 @@ class HttpUpstreamImpl : public HttpUpstream, protected Http::StreamCallbacks {
     }
 
     void decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override {
-      bool is_valid_response = Http::CodeUtility::is2xx(Http::Utility::getResponseStatus(*headers));
+      auto status = Http::Utility::getResponseStatus(*headers);
+      bool is_valid_post = parent_.tunnel_config_.usePost() && Http::CodeUtility::is2xx(status);
+      // In connect-udp case, Envoy will transform the H2 headers to H1 upgrade headers.
+      // A valid response should have SwitchingProtocol status and connect-udp upgrade.
+      bool is_valid_upgrade = status == 101 && Http::HeaderUtility::isConnectUdp(*headers);
+
+      ASSERT(!(is_valid_post && is_valid_upgrade));
+      bool is_valid_response = is_valid_post || is_valid_upgrade;
```

Comment:
Shouldn't we look at the config to see if we're in POST-mode instead? This will incorrectly accept some responses otherwise

Reply:
``is_valid_post`` can be true only if the config is set to ``usePost``, see line 305 above.
``is_valid_post`` and ``is_valid_upgrade`` can never be both true (asserted by line 310).
Not sure if I fully understood your comment, please lmk otherwise.

Reply:
Ah, you're right. Thanks for pointing it out. Could I propose to slightly refactor the code to make it easier to read? Perhaps:

```c++
      auto status = Http::Utility::getResponseStatus(*headers);
      bool is_valid_response;
      if (parent_.tunnel_config_.usePost()) {
        is_valid_response = Http::CodeUtility::is2xx(status);
      } else {
        // In connect-udp case, Envoy will transform the H2 headers to H1 upgrade headers.
        // A valid response should have SwitchingProtocol status and connect-udp upgrade.
        is_valid_response = status == 101 && Http::HeaderUtility::isConnectUdp(*headers);
      }
```

Reply:
Yeah looks better, thanks